### PR TITLE
Implementation Plan: Scoped Resume Picker (Issue #1064)

### LINKED ISSUE #1064: Scoped resume picker: filter cook vs order sessions with AutoSkillit-owned session browser
## Feature Description

Add an AutoSkillit-owned session picker that filters sessions by type (cook vs order) when the user runs `autoskillit cook --resume` or `autoskillit order --resume`. Instead of delegating entirely to Claude Code's unfiltered picker, AutoSkillit would read the JSONL session files directly, classify them, present a filtered picker, and pass the selected session ID to `claude --resume <id>`.

**Depends on:** #1063 (bare `--resume` bug fix) should land first. This feature builds on top of it.

## Motivation

With heavy use, the Claude Code resume picker shows a mixed list of cook sessions, order sessions, and potentially headless subagent sessions. Users who run `autoskillit cook --resume` likely want to see only cook sessions. Users who run `autoskillit order --resume` likely want to see only order sessions (ideally with the recipe name visible).

Claude Code's `--name` flag (the lighter alternative) has **known upstream persistence bugs** (#47197, #39657, #26249, #43963) that make name-based filtering unreliable. An AutoSkillit-owned picker avoids upstream coupling entirely.

## Prior Art

- **[cc-sessions](https://github.com/chronologos/cc-sessions)** — a third-party CLI that reads `~/.claude/projects/` JSONL files, extracts session metadata, and presents a filtered picker. Validates the JSONL parsing approach.
- **Claude Code's `--from-pr`** — an existing scoped-resume mechanism that filters sessions by PR association using `pr-link` entries in JSONL. Demonstrates the pattern of structured metadata → scoped resume.
- **Gemini CLI session management** — supports session listing with metadata display (title, timestamps, message count).
- **tmux session naming** — uses prefix conventions (`work_*`, `pers_*`) for session targeting.

### Session Storage Format

Sessions live at `~/.claude/projects/<encoded-cwd>/<session-uuid>.jsonl` where `<encoded-cwd>` is the absolute working directory with `/` and `_` replaced by `-`.

The `sessions-index.json` (per-project cache) has this schema:
```json
{
  "version": 1,
  "entries": [{
    "sessionId": "<uuid>",
    "fullPath": "<path-to-jsonl>",
    "fileMtime": 1768259664118,
    "firstPrompt": "...",
    "summary": "Optional short title",
    "messageCount": 15,
    "created": "2026-01-12T23:11:57.005Z",
    "modified": "2026-01-12T23:14:24.118Z",
    "gitBranch": "",
    "projectPath": "/home/...",
    "isSidechain": false
  }]
}
```

No `name`, `customTitle`, `tag`, or `sessionType` fields exist in this schema.

### How to Classify Cook vs Order Sessions

**Greeting heuristic (works today, zero structural changes needed):**

Order sessions always begin with one of 7 food-service greeting strings injected by the `order` command. Cook sessions never use these greetings.

`_COOK_GREETINGS` (4 entries, `src/autoskillit/cli/_prompts.py:68-76`):
```
"Welcome to Good Burger, home of the Good Burger, can I take your order? Today's special: {recipe_name}."
"Order up! Today's special: {recipe_name}. What ingredients are we working with?"
"Table for one! Today's special: {recipe_name}. Ready when you are."
"Fresh off the menu — today's special: {recipe_name}. What can I get started for you?"
```

`_OPEN_KITCHEN_GREETINGS` (3 entries, `src/autoskillit/cli/_prompts.py:78-82`):
```
"Welcome to Good Burger, home of the Good Burger, can I take your order?"
"Kitchen's open! What are we cooking today?"
"Order up! The kitchen is ready. What can I get you?"
```

**Classification rule:** A session is an `order` session if the first user message (the JSONL record with `type=user` and `parentUuid=null`) matches any greeting pattern. Otherwise it is a `cook` session.

**Heuristic reliability:**

| Scenario | Correctly classified? |
|---|---|---|
| Normal cook session | Yes — no greeting present |
| Cook with onboarding (first run) | Yes — first message is `/autoskillit:*` slash command |
| Normal order session | Yes — greeting always injected |
| Order with `--resume` | Yes — new greeting injected on each resume |
| Pre-greeting-era sessions (before 2026-03-13) | Yes — classified as cook (correct) |
| User manually types "Order up!" in cook | **False positive** — low probability but possible |

The greetings have been **completely stable** since introduction — 0 modifications across ~90 commits to `_prompts.py`.

**Stronger alternative — structural tagging (recommended for this implementation):**

Rather than relying solely on the greeting heuristic, add a structural session-type tag at launch time:

1. Inject `AUTOSKILLIT_SESSION_TYPE=cook|order` env var at launch via `env_extras` in `build_interactive_cmd`
2. Write a session registry entry at launch time to `.autoskillit/temp/session_registry.json`:
   ```json
   {
     "<ephemeral-session-id>": {
       "session_type": "cook",
       "launched_at": "2026-04-19T12:00:00Z",
       "recipe_name": null
     }
   }
   ```
3. On `open_kitchen` hook fire, update the registry with the Claude Code session UUID (bridging the two ID spaces — AutoSkillit's ephemeral `uuid4().hex[:16]` vs Claude Code's full UUID)
4. The picker reads the registry for classification; falls back to greeting heuristic for sessions predating the registry

The env var infrastructure is already in place: `build_interactive_cmd` accepts `env_extras`, the `_SESSION_BASELINE_ENV` pattern works, and `AUTOSKILLIT_SESSION_TYPE` would not be scrubbed by `build_claude_env` (confirmed by inspecting `_claude_env.py`'s denylist).

### Picker Implementation

**Input sources (in priority order):**
1. Session registry (`.autoskillit/temp/session_registry.json`) — structural, highest confidence
2. `sessions-index.json` — fast metadata without reading full JSONL files; provides `firstPrompt`, `summary`, `messageCount`, timestamps, `gitBranch`
3. Raw JSONL files — fallback for sessions not in registry; read first ~10 lines to find `type=user, parentUuid=null` record for greeting heuristic

**Display columns:**
- Session name/summary (from `sessions-index.json` `summary` field, or first prompt truncated)
- Type badge: `[cook]` / `[order: {recipe_name}]`
- Last modified timestamp
- Git branch
- Message count

**Picker library options:**
- `questionary` (already used by some Python CLIs; rich TUI selection)
- `fzf` subprocess (ubiquitous, fast, but external dependency)
- `simple_term_menu` (lightweight, no external deps)
- Custom using `_timed_input.py` patterns already in the codebase (no new dependency, but limited UX)

**Flow:**
```
autoskillit cook --resume
  → read session_registry.json (if exists)
  → read sessions-index.json for metadata
  → filter to cook sessions only
  → if 0 sessions: "No cook sessions found. Starting fresh."
  → if 1+ sessions: show picker
  → user selects session
  → subprocess: claude --resume <selected-uuid>
```

### New files
- Session registry writer (could live in `src/autoskillit/core/` or `src/autoskillit/workspace/`)
- Picker UI (could live in `src/autoskillit/cli/_session_picker.py`)

### Modified files
- `src/autoskillit/cli/_cook.py` — integrate picker into `cook(resume=True)` path
- `src/autoskillit/cli/app.py` — integrate picker into `order(resume=True)` path; write registry entry at launch
- `src/autoskillit/execution/commands.py` — pass `AUTOSKILLIT_SESSION_TYPE` env var
- `src/autoskillit/core/_type_enums.py` — (optional) add session type enum
- `src/autoskillit/core/paths.py` — extend `find_latest_session_id` with optional `session_type` filter, or add `claude_code_project_dir` as public API for the picker

### Potentially modified
- `src/autoskillit/hooks/open_kitchen_guard.py` — bridge Claude Code session UUID into registry
- `src/autoskillit/core/kitchen_state.py` — add `session_type` field to kitchen marker

## Key Design Decisions

1. **Picker library:** Evaluate `questionary` vs `simple_term_menu` vs custom. Custom using existing `_timed_input.py` patterns avoids new dependencies but has limited UX.
2. **Registry vs heuristic-only:** The session registry adds a write path at launch time but provides structural (non-textual) classification. The greeting heuristic works without the registry but is textual and has a (low-probability) false positive window.
3. **Fallback behavior:** When the picker finds no sessions of the requested type, should it fall back to showing all sessions or start a fresh session?
4. **`sessions-index.json` vs raw JSONL:** Reading the index is faster but the index may be stale or missing. Reading JSONL files is authoritative but slower for many sessions.

## Test Plan

- Unit tests for session classification (greeting heuristic + registry lookup)
- Unit tests for registry read/write
- Unit tests for picker filtering logic (mock JSONL/index data)
- Integration test: `cook --resume` with mixed session files shows only cook sessions
- Integration test: `order --resume` with mixed session files shows only order sessions
- Edge case: no sessions exist → graceful fallback
- Edge case: registry exists but is empty/corrupt → falls back to heuristic
- Edge case: pre-registry sessions → classified by greeting heuristic

## Dependencies

- **Blocked by:** #1194 (`order --resume <uuid>` UUID misroute + resume fast-path + system prompt duplication)
- **Blocks:** None
- **Why:** The scoped picker intercepts `BareResume` to show filtered sessions, then passes a selected UUID to `claude --resume <id>`. Three prerequisites from #1194 must land first: (1) the UUID misroute fix — without it, the picker-selected UUID routes to the `recipe` positional and fails; (2) the resume fast-path — without it, resuming without a recipe name triggers the full recipe validation gauntlet and errors out; (3) system prompt suppression on resume — without it, picker-resumed sessions get duplicate system prompts.

### PR DESCRIPTION
**Pull request by Trecek**
## Summary

Add an AutoSkillit-owned session picker that intercepts `BareResume` in `cook --resume`
and `order --resume` to show a filtered list of cook vs order sessions. The feature adds:

1. **`session_registry.py`** (L0 core, stdlib-only) — writes registry entries at launch
   with an ephemeral `AUTOSKILLIT_LAUNCH_ID`, later bridged to the Claude Code session UUID
   when `open_kitchen` fires.
2. **`_session_picker.py`** (L3 CLI) — reads sessions-index.json + registry, classifies
   by registry lookup or greeting heuristic, presents an interactive numbered picker.
3. **Registry bridge** in `open_kitchen_guard.py` — updates the registry entry with the
   Claude Code session UUID (bridging the two ID spaces).
4. **`BareResume` intercept** in `cook()` and `order()` — replaces the unfiltered Claude
   Code picker with the filtered AutoSkillit picker.

**Classification strategy:**
- Order sessions: structural (registry `claude_session_id` bridged via `open_kitchen` hook)
  with greeting-heuristic fallback for pre-registry sessions
- Cook sessions: no greeting → heuristic correctly identifies as cook
- Sessions with `isSidechain: true` are excluded from the picker

**Dependency note:** #1194 is NOT required for this path. `_run_interactive_session` already
suppresses `--append-system-prompt` for `NamedResume` (line 48), and the picker calls
`_launch_cook_session("", resume_spec=NamedResume(uuid))` which fast-paths past recipe
validation.

### State Lifecycle Diagram

```mermaid
%%{init: {'flowchart': {'nodeSpacing': 50, 'rankSpacing': 60, 'curve': 'basis'}}}%%
flowchart TB
    %% CLASS DEFINITIONS %%
    classDef cli fill:#1a237e,stroke:#7986cb,stroke-width:2px,color:#fff;
    classDef stateNode fill:#004d40,stroke:#4db6ac,stroke-width:2px,color:#fff;
    classDef handler fill:#e65100,stroke:#ffb74d,stroke-width:2px,color:#fff;
    classDef phase fill:#6a1b9a,stroke:#ba68c8,stroke-width:2px,color:#fff;
    classDef newComponent fill:#2e7d32,stroke:#81c784,stroke-width:2px,color:#fff;
    classDef output fill:#00695c,stroke:#4db6ac,stroke-width:2px,color:#fff;
    classDef detector fill:#b71c1c,stroke:#ef5350,stroke-width:2px,color:#fff;
    classDef gap fill:#ff6f00,stroke:#ffa726,stroke-width:2px,color:#000;
    classDef terminal fill:#1a237e,stroke:#7986cb,stroke-width:2px,color:#fff;

    START([CLI: cook --resume / order])

    subgraph ResumeSpec ["RESUMESPEC DISCRIMINATED UNION (frozen dataclasses)"]
        direction LR
        RS_NONE["NoResume<br/>━━━━━━━━━━<br/>INIT_ONLY<br/>Fresh session"]
        RS_BARE["● BareResume<br/>━━━━━━━━━━<br/>INIT_ONLY<br/>Delegate to picker"]
        RS_NAMED["NamedResume(session_id)<br/>━━━━━━━━━━<br/>INIT_ONLY<br/>Specific session ID"]
    end

    subgraph Registry ["★ SESSION REGISTRY (.autoskillit/temp/session_registry.json)"]
        direction TB
        INIT_FIELDS["INIT_ONLY Fields<br/>━━━━━━━━━━<br/>launch_id (uuid4 hex)<br/>session_type: cook | order<br/>launched_at (ISO8601)<br/>recipe_name (nullable)"]
        MUTABLE_FIELD["MUTABLE Field<br/>━━━━━━━━━━<br/>claude_session_id<br/>null → UUID<br/>(bridged by hook)"]
        ATOMIC_WRITE["★ _atomic_write()<br/>━━━━━━━━━━<br/>mkstemp + os.replace<br/>No partial writes"]
    end

    subgraph Picker ["★ SCOPED RESUME PICKER (_session_picker.py)"]
        direction TB
        LOAD_INDEX["_load_sessions_index()<br/>━━━━━━━━━━<br/>Filter: isSidechain=false<br/>READ-ONLY"]
        CLASSIFY["★ _classify_session()<br/>━━━━━━━━━━<br/>DERIVED — never stored"]
        TIER1_REG["Tier 1: Registry Lookup<br/>━━━━━━━━━━<br/>claude_session_id match"]
        TIER2_GREET["Tier 2: Greeting Heuristic<br/>━━━━━━━━━━<br/>firstPrompt prefix match<br/>→ 'order'"]
        TIER3_DEF["Tier 3: Default<br/>━━━━━━━━━━<br/>→ 'cook'"]
    end

    subgraph CookLoop ["● COOK SESSION LOOP (_cook.py)"]
        direction TB
        LAUNCH_ID["launch_id<br/>━━━━━━━━━━<br/>INIT_ONLY<br/>uuid4().hex[:16]"]
        WRITE_REG["● write_registry_entry()<br/>━━━━━━━━━━<br/>cook type<br/>claude_session_id=null"]
        RESUME_MUT["current_resume_spec<br/>━━━━━━━━━━<br/>MUTABLE: reload→NamedResume<br/>_first_run→False<br/>_initial_prompt→None"]
        RELOAD_SET["seen_reload_ids<br/>━━━━━━━━━━<br/>APPEND_ONLY<br/>Max 10 entries<br/>Dedup guard"]
    end

    subgraph OrderEntry ["● ORDER ENTRY (_session_launch.py)"]
        direction TB
        ORDER_LID["launch_id<br/>━━━━━━━━━━<br/>INIT_ONLY<br/>uuid4().hex[:16]"]
        WRITE_ORDER["● _write_order_entry()<br/>━━━━━━━━━━<br/>order type + recipe_name<br/>claude_session_id=null"]
    end

    subgraph KitchenGuard ["● OPEN_KITCHEN_GUARD (hook)"]
        direction TB
        HEADLESS_GATE["● Headless Gate<br/>━━━━━━━━━━<br/>fleet → DENY<br/>leaf/unset → DENY<br/>orchestrator → PERMIT"]
        KITCHEN_MARKER["Kitchen Marker (INIT_ONLY)<br/>━━━━━━━━━━<br/>session_id, opened_at<br/>recipe_name, marker_version<br/>Atomic write, never updated"]
        BRIDGE["★ _bridge_session_registry()<br/>━━━━━━━━━━<br/>AUTOSKILLIT_LAUNCH_ID → env<br/>Writes claude_session_id<br/>No-op if launch_id missing"]
    end

    %% FLOW %%
    START --> RS_NONE
    START --> RS_BARE
    START --> RS_NAMED

    RS_BARE --> LOAD_INDEX
    LOAD_INDEX --> CLASSIFY
    CLASSIFY --> TIER1_REG
    TIER1_REG -->|"match"| RS_NAMED
    TIER1_REG -->|"no match"| TIER2_GREET
    TIER2_GREET -->|"prefix match"| TIER3_DEF
    TIER2_GREET -->|"no match"| TIER3_DEF
    TIER3_DEF --> RS_NONE

    RS_NONE --> LAUNCH_ID
    RS_NAMED --> LAUNCH_ID
    LAUNCH_ID --> WRITE_REG
    WRITE_REG --> INIT_FIELDS
    INIT_FIELDS --> MUTABLE_FIELD
    MUTABLE_FIELD --> ATOMIC_WRITE

    LAUNCH_ID --> ORDER_LID
    ORDER_LID --> WRITE_ORDER
    WRITE_ORDER --> INIT_FIELDS

    WRITE_REG --> RESUME_MUT
    RESUME_MUT --> RELOAD_SET
    RELOAD_SET -->|"reload detected"| RESUME_MUT

    WRITE_REG --> HEADLESS_GATE
    HEADLESS_GATE -->|"permit"| KITCHEN_MARKER
    KITCHEN_MARKER --> BRIDGE
    BRIDGE --> MUTABLE_FIELD

    END_NODE([Session Active])

    KITCHEN_MARKER --> END_NODE
    RELOAD_SET -->|"None / max reached"| END_NODE

    %% CLASS ASSIGNMENTS %%
    class START,END_NODE terminal;
    class RS_NONE,RS_BARE,RS_NAMED cli;
    class LOAD_INDEX,CLASSIFY,TIER1_REG,TIER2_GREET,TIER3_DEF newComponent;
    class INIT_FIELDS,MUTABLE_FIELD stateNode;
    class ATOMIC_WRITE,BRIDGE newComponent;
    class LAUNCH_ID,ORDER_LID detector;
    class WRITE_REG,WRITE_ORDER handler;
    class RESUME_MUT phase;
    class RELOAD_SET gap;
    class HEADLESS_GATE detector;
    class KITCHEN_MARKER output;
```

Closes #1064

## Implementation Plan

Plan file: `/home/talon/projects/autoskillit-runs/impl-20260425-213021-965509/.autoskillit/temp/make-plan/scoped_resume_picker_plan_2026-04-25_120000.md`

🤖 Generated with [Claude Code](https://claude.com/claude-code) via AutoSkillit

## Token Usage Summary

| Step | uncached | output | cache_read | cache_write | count | time |
|------|----------|--------|------------|-------------|-------|------|
| plan | 189 | 32.0k | 1.6M | 122.7k | 1 | 9m 24s |
| verify | 92 | 11.5k | 445.9k | 43.4k | 1 | 4m 8s |
| implement | 816 | 63.9k | 9.9M | 151.9k | 1 | 22m 9s |
| fix | 1.0k | 118.8k | 13.4M | 250.6k | 2 | 41m 28s |
| prepare_pr | 84 | 6.8k | 326.3k | 32.9k | 1 | 1m 57s |
| run_arch_lenses | 77 | 5.9k | 284.0k | 38.7k | 1 | 1m 45s |
| compose_pr | 52 | 4.1k | 162.6k | 24.5k | 1 | 1m 23s |
| **Total** | 2.3k | 243.1k | 26.1M | 664.7k | | 1h 22m |

### CODE REVIEW

File: src/autoskillit/hooks/open_kitchen_guard.py [hunk 1]
```diff
@@ -53,6 +53,40 @@ def _write_kitchen_marker(session_id: str, recipe_name: str | None) -> None:
         raise
 
 
+def _bridge_session_registry(session_id: str) -> None:
```
Comment (Trecek) [inline src/autoskillit/hooks/open_kitchen_guard.py:56]: [warning] cohesion: `_bridge_session_registry` re-implements the atomic-write and registry-update logic already in `session_registry.bridge_claude_session_id`. Since `session_registry.py` is stdlib-only, the hook can import it directly: `from autoskillit.core.session_registry import bridge_claude_session_id` and call `bridge_claude_session_id(Path.cwd(), launch_id, session_id)` — eliminates ~25 lines of duplicated atomic write code.

File: src/autoskillit/hooks/open_kitchen_guard.py [hunk 2]
```diff
@@ -53,6 +53,40 @@ def _write_kitchen_marker(session_id: str, recipe_name: str | None) -> None:
...
+    registry[launch_id]["claude_session_id"] = session_id
+
+    registry_file.parent.mkdir(parents=True, exist_ok=True)
+    content = json.dumps(registry)
+    fd, tmp = tempfile.mkstemp(dir=registry_file.parent, suffix=".tmp")
+    try:
+        with os.fdopen(fd, "w", encoding="utf-8") as f:
+            f.write(content)
+        os.replace(tmp, registry_file)
+    except Exception:
```
Comment (Trecek) [inline src/autoskillit/hooks/open_kitchen_guard.py:83]: [warning] defense: The `except Exception` cleanup block never re-raises, so atomic write failures are silently swallowed. The caller's try/except error-reporting block at line 147 is dead code for all write failures. Either add `raise` after `os.unlink(tmp)` cleanup, or remove the caller's try/except since it can never fire for write errors.

File: src/autoskillit/core/__init__.py [hunk 3]
```diff
@@ -94,6 +94,18 @@
     readiness_sentinel_path,
     write_readiness_sentinel,
 )
+from .session_registry import (
```
Comment (Trecek) [inline src/autoskillit/core/__init__.py:97]: [warning] cohesion: session_registry re-exports are split into 4 separate single-item import blocks. Every other module in this file groups all its re-exports into one block (e.g., readiness, kitchen_state). Consolidate into a single `from .session_registry import (bridge_claude_session_id as bridge_claude_session_id, read_registry as read_registry, ...)` block.

File: src/autoskillit/cli/_session_launch.py [hunk 4]
```diff
@@ -57,6 +57,22 @@ def _run_interactive_session(
     return None
 
 
+def _write_order_entry(project_dir: Path, recipe_name: str | None) -> dict[str, str]:
+    """Generate an order launch ID, write registry entry, return env extras dict."""
+    import uuid
```
Comment (Trecek) [inline src/autoskillit/cli/_session_launch.py:61]: [info] slop: Docstring `"Generate an order launch ID, write registry entry, return env extras dict."` describes WHAT the function does (mechanics) rather than WHY — violates project convention against what-comments. Remove it.

**Review by Trecek — COMMENTED** [review]
AutoSkillit PR Review — Verdict: approved_with_comments

**Review by Trecek — COMMENTED** [review]
AutoSkillit review: warning-only findings detected. See inline comments — no blocking changes required.

File: src/autoskillit/hooks/open_kitchen_guard.py (hunk 1 above)
Comment (Trecek) [reply src/autoskillit/hooks/open_kitchen_guard.py:56]: Investigated — this is intentional. open_kitchen_guard.py is a stdlib-only hook enforced by test_hooks_are_stdlib_only. Importing from autoskillit.core.session_registry would violate that constraint — the inline duplication is by design. See _hook_settings.py:12.

File: src/autoskillit/hooks/open_kitchen_guard.py (hunk 2 above)
Comment (Trecek) [reply src/autoskillit/hooks/open_kitchen_guard.py:83]: Agreed — fixed in dedff7c0. Added raise to the cleanup block so atomic write failures propagate to the caller stderr logger, consistent with _atomic_write in session_registry.py and _write_kitchen_marker in the same file.

File: src/autoskillit/core/__init__.py (hunk 3 above)
Comment (Trecek) [reply src/autoskillit/core/__init__.py:97]: Investigated — this is intentional. ruff I001 (isort) enforces single-item blocks for as-aliased re-exports in this file. Consolidating triggers a fixable I001 violation; ruff --fix immediately expands them back. The 4-block format is ruff-mandated.

File: src/autoskillit/cli/_session_launch.py (hunk 4 above)
Comment (Trecek) [reply src/autoskillit/cli/_session_launch.py:61]: Agreed — fixed in 54e56476. Removed the docstring that described mechanics rather than intent. Function name _write_order_entry already communicates the operation.
